docker/pusher.go

```diff
@@ -27,6 +31,92 @@ type Credentials struct {
 
 type CredentialsProvider func(ctx context.Context, registry string) (Credentials, error)
 
+type CredentialsCallback func(registry string) (Credentials, error)
+
+var ErrUnauthorized = errors.New("bad credentials")
+
+// VerifyCredentialsCallback checks if credentials are accepted by the registry.
+// If credentials are incorrect this callback shall return ErrUnauthorized.
+type VerifyCredentialsCallback func(ctx context.Context, username, password, registry string) error
+
+func CheckAuth(ctx context.Context, username, password, registry string) error {
+	err := dc.CheckAuth(ctx, nil, username, password, registry)
+	unAuthErr := dc.ErrUnauthorizedForCredentials{}
+	if errors.As(err, &unAuthErr) {
+		return ErrUnauthorized
+	}
+	return err
+}
+
+// NewCredentialsProvider returns new CredentialsProvider that tires to read credentials from `~/.docker/config.json`.
+// If reading credentials from the config fails caller provided callback will be invoked to obtain credentials.
+// The callback will typically prompt user to enter password to stdin.
+// To verify that password is correct verifyCredentials param may be used.
+// If verifyCredentials == nil then CheckAuth will be used.
+func NewCredentialsProvider(credentialsCallback CredentialsCallback, verifyCredentials VerifyCredentialsCallback) CredentialsProvider {
+	if verifyCredentials == nil {
+		verifyCredentials = CheckAuth
+	}
+	return func(ctx context.Context, registry string) (Credentials, error) {
+
+		result := Credentials{}
+		credentials, err := config.GetCredentials(nil, registry)
+		if err != nil {
+			return result, errors.Wrap(err, "failed to get credentials")
+		}
+
+		if credentials != (containersTypes.DockerAuthConfig{}) {
+			result.Username, result.Password = credentials.Username, credentials.Password
+
+			err = verifyCredentials(ctx, result.Username, result.Password, registry)
+			if err == nil {
+				return result, nil
+			} else {
+				if !errors.Is(err, ErrUnauthorized) {
+					return Credentials{}, err
+				}
+			}
+		}
+
+		credentials, err = GetCredentialsFromCredsStore(registry)
+		if err != nil {
+			return Credentials{}, err
+		}
+
+		if credentials != (containersTypes.DockerAuthConfig{}) {
+			result.Username, result.Password = credentials.Username, credentials.Password
+
+			err = verifyCredentials(ctx, result.Username, result.Password, registry)
+			if err == nil {
+				return result, nil
+			} else {
+				if !errors.Is(err, ErrUnauthorized) {
+					return Credentials{}, err
+				}
+			}
+		}
+
+		for {
+			result, err = credentialsCallback(registry)
+			if err != nil {
+				return Credentials{}, err
+			}
+
+			err = verifyCredentials(ctx, result.Username, result.Password, registry)
+			if err == nil {
+				// TODO maybe save the credentials
```

Comment:
@zroubalik right now we are only reading `~/.docker/config.json` and we are never writing it. I think we shouldn't modify it but rather create our own file e.g. `~/.config/func/auth.json`. Question is: if we will introduce such a file should we continue to read also `~/.docker/config.json`, or maybe just one time import it first time?

Reply:
Hello @matejvasek I think if the new file get introduced, func could still read `~/.docker/config.json` once users (along the path) may decidelogin or switch to different registries after first-time import, using docker commands or some specific auth input method depending the registry (i.e. gcr.io with gcloud auth). In this case auth creadentials for such registry can be combined with docker/podman login commands on CI systems.

Reply:
I think save is a good idea, otherwise it will keep asking for user/pw every time.

Reply:
It might be nice to import the values from the docker config file at least on the first run, but that's not even necessary if we are caching the credentials.  It definitely makes sense for us not to start depending on / writing the config of other systems and instead use a config specific to func such as `~/.config/func/auth.json`